packages/nimble-components/src/utilities/models/array-to-tree.ts

```diff
@@ -0,0 +1,121 @@
+// Modified from https://github.com/philipstanislaus/performant-array-to-tree/blob/v1.11.0/src/arrayToTree.ts
+// Copyright (c) 2022 philipstanislaus
+// SPDX-License-Identifier: MIT
+
+export interface TreeItem<Item> {
+    clientRecord: Item | undefined;
+    subRows: TreeItem<Item>[];
+    originalIndex: number | undefined;
+}
```

Comment:
Why not just return a TreeNode as defined in table/types? Can then avoid the `as TreeNode` cast in the table code

Reply:
This might require a redesign of the algorithm which requires setting a couple of these fields to `undefined` in a temporary state, which `TableNode` at the moment does not allow (and I think would be awkward if it did). For now I would prefer to leave as-is, unless you have an suggestion I'm not seeing to resolve this.

Reply:
You can have a type used in the file as, `type TreeNodePartial = Partial<TreeNode>` and then at return do `return rootItems as TreeNode`.

Reply:
Mainly have that connection to the TreeNode type of what fields are defined and encapsulate the knowledge that arrayToTree satisfies the TreeNode contract in the method (instead of the callers of the method having to assert that)